### ISSUE TITLE
Define custom attribute with `def`

### ISSUE THREAD

**Issue by danielmalaton** [okuramasafumi/alba#277]
## Describe the bug
Define custom attribute with [`def` method](https://github.com/okuramasafumi/alba#simple-serialization-with-root-key) which method's name is the same as one of the models will return unexpectedly. 

## To Reproduce
Suppose in `schema.rb` has table `a` and table `b`. **`a` belongs to `b`**
In class `AResource`
```Ruby
class AResource
  include Alba::Resource

  attributes :id, :name, :b
  
  def b(_)
    true
  end
end
```

## Expected behavior
When serializing an AResource object, it should return
```JSON
[{ "id": 1, "name": "abc", "b": true }, { "id": 2, "name": "xyz", "b": true }]
```

## Actual behavior
In `b` attribute, it does not return anything (or it queries `b` instance object if it had `b_id`)

## Environment
- Ruby version: 2.7.4

## Additional context
Temporary:
```Ruby
class AResource
  include Alba::Resource

  attributes :id, :name 
  attribute :b do 
    b
  end
  
  def b
    true
  end
end
``` 
will work. But I don't think it is a best practice.

**Comment by okuramasafumi** [comment]
@danielmalaton What do you want to achieve by defining a method with the same name as an existing attribute? In this case, why would you want to return `true` with the same name as an association?

Alba first tries to fetch attribute from a given object, and if it cannot find it then tries to find a method defined in a resource. 
I think your temporal solution would be fine with some comments. If you find it not, then what would be the best solution?

**Comment by danielmalaton** [comment]
Updated: Entity `a` must belong to entity `b`.

I tried to define custom attributes called `b` in `AResource`. But if I go with `attribute`, it will throw [Sorbet](https://github.com/chanzuckerberg/sorbet-rails) error: Method `params` does not exist on `T.class_of(AResource)` (fix available)[7003](https://srb.help/7003)
```Ruby
attribute :b do |object|
  b = object.b
  # play with params
end
```
Then I decide to go with `def` to define the custom attributes. But in this case
```Ruby
def b(_)
  # magic logic
end
```
It will not return expectly, it returns `object.b` instead. What should I do in this case?

**Comment by okuramasafumi** [comment]
Alba doesn't support Sorbet. Solve your problem on your own (or simply ignore it).

If `a` and `b` both are something like `ActiveRecord::Base` object, you should use `association` (or its alias, `one`, for example) to serialize `b` and embed the result in `a`.

**Comment by datbth** [comment]
hi, I'd like to share a bit more on why I think it is more **convenient** to make Alba prioritize the methods over the object attributes with the same names.  

Take this current code for example:
```ruby
class UserResource
  include Alba::Resource
  attributes :id, :name
  
  attribute :role do
    normalized_role
  end
  
  attribute :permissions do |object|
    {
       view: object.permissions & 0x01,
       edit: object.permissions & 0x02,
       manage: normalized_role == 'super_admin'
    }
  end

  private

  def normalized_role
    @normalized_role ||=
      if object.role == 'admin' && object.created_at > '2023-01-01' && costly_check(object)
        'super_admin'
      else
        object.role
      end
  end
end
```

A few notes:
* `User` has a `permissions` column that needs to be serialized differently when responding to clients
* The `role` column also needs some extra processing due to business changes. The processing is quite costly that it would benefit from a simple caching.
  * I know that there are other solutions such as normalizing the database data, normalizing/caching before calling alba, etc. But they are not convenient and might not be feasible in complex systems.

With Sorbet, it would look like this:
```ruby
class UserResource
  extend T::Sig
  include Alba::Resource

  attributes :id, :name
  
  attribute :role do
    T.bind(self, UserResource)
    normalized_role
  end
  
  attribute :permissions do |object|
    T.bind(self, UserResource)
    {
       view: object.permissions & 0x01,
       edit: object.permissions & 0x02,
       manage: normalized_role == 'super_admin'
    }
  end

  private

  sig { returns(String) }
  def normalized_role
    @normalized_role ||=
      if object.role == 'admin' && object.created_at > '2023-01-01' && UserPayment.costly_check(object)
        'super_admin'
      else
        object.role
      end
  end
end
```

This is quite verbose.  
Using procs instead of methods would make it less verbose, but then I would have to compromise the caching.

If Alba prioritizes methods over object attributes, it could be written like this:
```ruby
class UserResource
  extend T::Sig
  include Alba::Resource

  attributes :id, :name, :role, :permissions
  
  sig { returns(String) }
  def role
    @role ||=
      if object.role == 'admin' && object.created_at > '2023-01-01' && UserPayment.costly_check(object)
        'super_admin'
      else
        object.role
      end
  end

  sig { returns(T::Hash[Symbol, T::Boolean]) }
  def permissions 
    {
       view: object.permissions & 0x01,
       edit: object.permissions & 0x02,
       manage: role == 'super_admin'
    }
  end
end
```
This is more convenient because:
* I don't need to make new names/variables for `role`
* I can still benefit from Sorbet without much overhead
* It is similar to ActiveRecord Serializers, whose behavior I'm already familiar with

On the other hand, I could not really come up with use cases where I would need to define methods having the same names with the object attributes, while still returning the original object attributes. Could you share if you have such use cases, or is this more of a technical decision where the current behavior is easier to implement and maintain?

To sum up, Alba is still working fine and it is not blocking any use case. I only think it would be more **convenient** to prioritize methods over object attributes with the same names, so I'd appreciate you considering this behavioral change.

**Comment by okuramasafumi** [comment]
> Could you share if you have such use cases, or is this more of a technical decision where the current behavior is more performant, or is easier to implement and maintain?

The answer is because Alba started to support method attribute later than supporting block attribute. I just "added" method attribute implementation, so Alba behaves as now.

It's possible to change current behavior, but it's not backward compatible. We can add some flag such as `Alba.prefer_method_attribute` for migration with `false` default value, and make the default as `true` in the next major version.

Btw, I think it's the problem of Sorbet that it's hard to type DSLs. I don't want to write `T.bind` lines so many times. I hope Sorbet will have some ways to type DSLs more easily.

**Comment by okuramasafumi** [comment]
@danielmalaton @datbth 
So here's the code for `attributes` DSL to work.

```ruby
      def fetch_attribute_from_object_and_resource(obj, attribute)
        obj.__send__(attribute)
      rescue NoMethodError
        __send__(attribute, obj)
      end
```

As you can see, it first sends a message to underlying object, and if it fails, it sends the same message to itself.
This is just a method of a resource, so you can override it like below:

```ruby
      def fetch_attribute_from_object_and_resource(obj, attribute)
        __send__(attribute, obj)
      rescue NoMethodError
        obj.__send__(attribute)
      end
```

Note that we don't use `respond_to?` because it's relatively slow.

This is a quick fix you can do today. I know it's not the best way, so I'll add some flag in the future.

**Comment by trevorturk** [comment]
I wanted to chime in on this thread to show another use case I have, where I want to "override" an `association` object/method. 

The details aren't super important, but just to illustrate, here I'm trying to limit an array of items called `data` so it always returns a multiple of 8. I can do this inline like so:

```ruby
association :data, proc { |data, _, _| data.in_groups_of(8).reject { |d| d.any?(&:nil?) }.flatten } do
  attributes :temperature
end
```

...but I wonder if there's a better way, or if this might be helpful if/when considering adding a flag in the future so it might cover this use case as well.
